Suggestion: do not describe all properties twice

make-it-git opened the pull request:
May be it is enough to describe them only in $fields. $fields are self explanatory

make-it-git committed:
Suggestion: do not describe all properties twice

May be it is enough to describe them only in $fields

make-it-git committed:
Merge branch 'dev' of github.com:darrynten/sage-one-php into dev

make-it-git committed:
Merge branch 'dev' into remove-field-duplicates

make-it-git committed:
Trying to restore code coverage

TODO Update AccountModelTest to reduce duplicate code

darrynten commented on src/BaseModel.php:
```diff
@@ -109,22 +116,33 @@ public function __set($key, $value)
+                // so we return it as default value for any described but not loaded property
+                return null;
+            }
+            return $this->fieldsData[$key];
```
We can simplify, and flatten this a bit and reduce complexity

```php
public function __get($key)
{
    if (array_key_exists($key, $this->fields)) {
        if (array_key_exists('default', $this->fields[$key])) {
            // we can leave early
            return $this->fields[$key]['default'];
        }

        if (array_key_exists($key, $this->fieldsData)) {
            return $this->fieldsData[$key];
        }

        return null;
    }

    // normal exception handler
```

I have not tested but it should work

darrynten reviewed (changes requested):
Reducing complexity in __get() suggestion

darrynten commented on src/BaseModel.php:
```diff
@@ -258,24 +278,26 @@ private function getRemoteKey($localKey)
      */
     private function prepareObjectRow($key, $config)
     {
+        $value = $this->__get($key);
```
Good call this 🍺

make-it-git replied:
Is there better way?

darrynten replied:
I think that's the right way

darrynten commented on tests/SageOne/Models/AccountModelTest.php:
```diff
@@ -343,7 +332,10 @@ public function testSave()
+
+        $response = $accountModel->save();
+        $this->assertEquals($expectedResponse->ID, $response->ID);
         // (var_dump($accountModel->toJson()));
```
We can't have commented out code getting committed ;)

darrynten commented on tests/SageOne/Models/ExampleModelTest.php:
```diff
@@ -357,9 +353,6 @@ public function testExamples()
 
-        // Expected lengths after loading
-        $this->assertCount(8, (array)$exampleModel);
-
```
Any particular reason for these assertions to not still be there?

make-it-git replied:
This test will be rewritten with new methods, there is already this check.
This one checks not only fields, but also other properties

make-it-git committed:
Added tests for default values

darrynten commented on src/BaseModel.php:
```diff
@@ -109,22 +116,31 @@ public function __set($key, $value)
+                // there is some data loaded so we return it
+                return $this->fieldsData[$key];
+            }
+            if (array_key_exists('default', $this->fields[$key])) {
```
Where does the `default` key come from?

make-it-git replied:
I guess from there docs.
If key is nullable, we are allowed not to specify it because it is null by default

make-it-git replied:
May be this is not required.
May be throw exception in __get if data is not loaded?

darrynten replied:
So it's maybe that we have potential `default` and `required` properties to add to the fields map?

darrynten committed:
Additional example fields

darrynten committed:
Replaced 'persistable' with 'readonly'

Read only makes more sense than persistable as it is what the
official documentation uses.

Inverted all appropriate booleans.

darrynten committed:
Errant file

darrynten committed:
Formatting of tests

darrynten committed:
Fixed test true/false attribute check

darrynten committed:
Updated BaseModelTest docblock indicating attributes are all mandatory

darrynten committed:
Consistent string/variable interpolation

darrynten committed:
Match assertion strings

darrynten committed:
Updated example mock to match expectations

darrynten committed:
Added validation and tests for min/max and regex optional field params

darrynten committed:
More defensive test examples

make-it-git committed:
Merge branch 'dev' of github.com:darrynten/sage-one-php into dev

darrynten committed:
Merge branch 'dev' into remove-field-duplicates-darryn

* dev:
  Mocks for various HTTP Response Codes
  Added github template for Issues and Pull Requests (#18)

darrynten committed:
Use multibyte string length checks

darrynten committed:
Errant newline

make-it-git committed:
Merge remote-tracking branch 'remotes/origin/remove-field-duplicates-darryn' into remove-field-duplicates

make-it-git committed:
Merge branch 'dev' into remove-field-duplicates

make-it-git committed:
Modified Account model tests

darrynten commented on tests/SageOne/Models/BaseModelTest.php:
```diff
@@ -189,23 +182,32 @@ protected function verifyAttributes(string $class, array $attributes)
+            $readonly = $options['readonly'];
+            $readonlyText = $readonly ? 'true' : 'false';
+            $readonlyOptionText = $value[$name]['readonly'] ? 'true' : 'false';
+
```
I think we add additional checks on min/max/regex etc here

darrynten commented on src/Models/AccountCategory.php:
```diff
@@ -99,32 +53,32 @@ class AccountCategory extends BaseModel
             'type' => 'integer',
             'nullable' => false,
-            'persistable' => true,
+            'readonly' => false,
```
This now has a `required` on it if we look at the docs and our new rules

https://accounting.sageone.co.za/api/1.1.2/Help/ResourceModel?modelName=AccountCategory

darrynten reviewed (changes requested):
So far so good, very minor requests

darrynten committed:
Moved validation to helper class

make-it-git committed:
Merge pull request #22 from darrynten/ref-base-model

Moved validation to helper class

darrynten committed:
Shortened long variable name

make-it-git committed:
Merge branch 'remove-field-duplicates' of github.com:darrynten/sage-one-php into remove-field-duplicates

make-it-git committed:
Use Trait in order to resolve codeclimate complaints

make-it-git committed:
StyleCI fix

darrynten committed:
AccountCategory field id is `required`
